File: web-platform/types/oracle.ts
/**
 * Oracle 2.0 TypeScript interfaces
 */

export interface OracleSource {
  type: 'REPORT' | 'ARTICOLO';
  id: number | null;
  title: string;
  date_str?: string;
  similarity: number;
  status?: string;
  preview?: string;
  link?: string;
  source?: string;
}

export interface ExecutionStep {
  tool_name: string;
  parameters: Record<string, unknown>;
  is_critical: boolean;
  description: string;
}

export interface QueryPlan {
  intent: 'factual' | 'analytical' | 'narrative' | 'market' | 'comparative' | 'overview';
  complexity: 'simple' | 'medium' | 'complex';
  tools: string[];
  execution_steps: ExecutionStep[];
  estimated_time: number;
  requires_decomposition: boolean;
  sub_queries?: string[];
}

export interface OracleResponse {
  answer: string;
  sources: OracleSource[];
  query_plan?: QueryPlan;
  mode: string;
  metadata: {
    query: string;
    session_id: string;
    is_follow_up: boolean;
    execution_time: number;
    tools_executed: string[];
    timestamp: string;
  };
}

export interface OracleChatMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  timestamp: string;
  sources?: OracleSource[];
  query_plan?: QueryPlan;
  metadata?: Record<string, unknown>;
}

export interface OracleChatFilters {
  mode?: 'both' | 'factual' | 'strategic';
  search_type?: 'vector' | 'keyword' | 'hybrid';
  start_date?: string;
  end_date?: string;
  categories?: string[];
  gpe_filter?: string[];
}

export interface OracleActiveFilters {
  mode?: 'both' | 'factual' | 'strategic';
  search_type?: 'vector' | 'keyword' | 'hybrid';
  start_date?: string;
  end_date?: string;
  gpe_filter?: string[];
}

File: src/api/schemas/oracle.py
"""Pydantic schemas for Oracle 2.0 API endpoints."""

import re
from datetime import date
from typing import Any, Annotated, Dict, List, Literal, Optional

from pydantic import BaseModel, Field, field_validator

# Alphanumeric + hyphens/underscores only — prevents session ID injection
_SESSION_ID_RE = re.compile(r"^[a-zA-Z0-9_\-]{1,64}$")

# Strip null bytes from query strings
_NULL_BYTE_RE = re.compile(r"\x00")


class OracleChatRequest(BaseModel):
    query: str = Field(..., min_length=3, max_length=2000)
    session_id: str = Field(default="default", max_length=64)
    mode: Literal["both", "factual", "strategic"] = "both"
    search_type: Literal["vector", "keyword", "hybrid"] = "hybrid"
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    categories: Optional[List[Literal[
        "GEOPOLITICS", "DEFENSE", "ECONOMY", "CYBER", "ENERGY"
    ]]] = None
    gpe_filter: Optional[List[Annotated[str, Field(max_length=100)]]] = Field(default=None, max_length=10)
    # BREAKING CHANGE (2026-04-17): gemini_api_key field removed.
    # Oracle now uses server-side ANTHROPIC_API_KEY. Passing this field returns HTTP 422.

    @field_validator("query")
    @classmethod
    def sanitize_query(cls, v: str) -> str:
        return _NULL_BYTE_RE.sub("", v).strip()

    @field_validator("session_id")
    @classmethod
    def validate_session_id(cls, v: str) -> str:
        if not _SESSION_ID_RE.match(v):
            raise ValueError("session_id must be alphanumeric with hyphens/underscores only")
        return v


class OracleSource(BaseModel):
    type: Literal["REPORT", "ARTICOLO"]
    id: Optional[int] = None
    title: str
    date_str: Optional[str] = None
    similarity: float = 0.0
    status: Optional[str] = None
    preview: Optional[str] = None
    link: Optional[str] = None
    source: Optional[str] = None


class OracleChatResponse(BaseModel):
    answer: str
    sources: List[OracleSource] = []
    query_plan: Optional[Dict[str, Any]] = None
    mode: str = "both"
    metadata: Dict[str, Any] = {}
